### PR TITLE
JS Asset Auditor engineering spec

### LINKED ISSUE #606: JS Asset Auditor — Publisher Page Sweep and Config Generation
As a publisher I would like to be able to run an audit of my web properties to discover all third-party JS assets currently loading on my pages. The audit should automatically generate a configuration file that maps each asset to a first-party proxy path, detect which assets load in the page head, and allow me to re-run the audit at any time to surface new or removed assets.

### LINKED ISSUE #607: Create Spec for 3rd Party JS Asset Auditor
### Description

Create Spec for 3rd Party JS Asset Auditor

### PR DESCRIPTION
**Pull request by jevansnyc**
## Summary

Engineering spec for the JS Asset Auditor — a Claude Code slash command that sweeps a publisher page via Chrome DevTools MCP, detects third-party JS assets, and generates or diffs `js-assets.toml`.

Closes #606

## Docs added

- `docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md` — engineering spec covering sweep protocol, Chrome DevTools MCP tooling, heuristic filtering, slug generation, init and diff modes

## Test plan

- [ ] Review sweep protocol and verify Chrome DevTools MCP tools match what is available
- [ ] Verify heuristic filter list covers expected ad tech vs non-ad tech origins
- [ ] Confirm slug algorithm matches proxy spec derivation

### CODE REVIEW

**Review by prk-Jr — COMMENTED** [review]
## Summary

Reviewed the engineering spec and left a few implementation-facing questions inline. The main areas are URL normalization, wildcarding, third-party boundary matching, path derivation, and `inject_in_head` semantics for dynamic scripts.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+---
+
+## Sweep Protocol
+
+1. Read `trusted-server.toml` → extract `publisher.domain` (defines first-party boundary)
+2. Open Chrome via `mcp__chrome-devtools__new_page`, navigate to target URL via `mcp__chrome-devtools__navigate_page`
+3. Wait for full page load + ~6s settle window for async script loads (`mcp__chrome-devtools__wait_for`)
+4. In parallel:
+   - `mcp__chrome-devtools__list_network_requests` → filter for requests where URL ends in `.js` or `Content-Type: application/javascript`, and origin ≠ `publisher.domain`
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🔧 **wrench** — MCP tool names are wrong throughout the spec

Every `mcp__chrome-devtools__*` reference uses an incorrect prefix. The actual tool names in `.claude/settings.json` and `.claude/settings.local.json` are:

```
mcp__plugin_chrome-devtools-mcp_chrome-devtools__new_page
mcp__plugin_chrome-devtools-mcp_chrome-devtools__navigate_page
mcp__plugin_chrome-devtools-mcp_chrome-devtools__list_network_requests
...etc
```

Affects 12 occurrences across the Sweep Protocol (lines 36-43) and Implementation (lines 188-193) sections. A skill implemented from this spec will fail on every MCP call.

**Fix:** Replace all `mcp__chrome-devtools__` with `mcp__plugin_chrome-devtools-mcp_chrome-devtools__`.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+   - `mcp__chrome-devtools__evaluate_script` → `Array.from(document.head.querySelectorAll('script[src]')).map(s => s.src)` → collect head-loaded script URLs
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🔧 **wrench** — `wait_for` does not provide a settle window

The spec says: _"Wait for full page load + ~6s settle window for async script loads (`wait_for`)"_

But `wait_for` waits for **specific text to appear on the page** (params: `text` — list of strings, `timeout` — max wait ms). It is not a network-idle or timeout mechanism.

**Fix:** Use `evaluate_script` with:
```js
await new Promise(r => setTimeout(r, 6000))
```
Or poll `list_network_requests` to detect when no new script requests appear.

---

🤔 **thinking** — 6s may be too short for complex ad tech pages

Header bidding waterfalls, consent-gated loading, and lazy demand partners can take 10-15s. Consider making this configurable (e.g., `--settle 10s`) or documenting the tradeoff.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+5. Apply heuristic filter (see below)
+6. For each surviving asset, generate a `[[js_assets]]` entry (see below)
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🔧 **wrench** — `list_network_requests` filtering claim is inaccurate

The spec says: _"filter for requests where URL ends in `.js` or `Content-Type: application/javascript`"_

The tool supports `resourceTypes` filtering (e.g., `["script"]`) but does **not** support URL pattern or Content-Type filtering. Those must be done in post-processing.

**Fix:** Rewrite to: _"`list_network_requests` with `resourceTypes: ["script"]` → post-filter to exclude first-party origins matching `publisher.domain`"_

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+| Category | Excluded origins |
+|---|---|
+| Framework CDNs | `cdnjs.cloudflare.com`, `ajax.googleapis.com`, `cdn.jsdelivr.net`, `unpkg.com` |
+| Error tracking | `sentry.io`, `bugsnag.com`, `rollbar.com` |
+| Font services | `fonts.googleapis.com`, `fonts.gstatic.com` |
+| Social embeds | `platform.twitter.com`, `connect.facebook.net` |
+
+**`googletagmanager.com` is not filtered** — GTM is ad tech and should be proxied.
+
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🤔 **thinking** — `platform.twitter.com` may be outdated

Twitter rebranded to X. The domain may already be `platform.x.com` or will change. Consider listing both, or verifying the current domain.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+Everything else surfaces for operator review.
+
+---
+
+## Asset Entry Generation
+
+| Field | Derivation |
+|---|---|
+| `slug` | `{publisher_prefix}:{asset_stem}` — see slug algorithm below |
+| `path` | `/{publisher_prefix}/{asset_stem}.js`, or wildcard variant if versioned path detected |
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🔧 **wrench** — Path derivation algorithm contradicts the examples

This table says `path = "/{publisher_prefix}/{asset_stem}.js"`. For the first example (`publisher_prefix = aB3kR7mN`, `asset_stem = prebid-load`), this yields `/aB3kR7mN/prebid-load.js`.

But the example on line 119 shows `path = "/sdk/aB3kR7mN.js"` — a completely different structure. The Proxy spec shows the same contradictory examples.

Since `path` is the routing key between Auditor and Proxy, this must be resolved.

**Fix:** Align the algorithm description with the examples (or vice versa).

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+**Rationale:** Fully opaque and hash-derived — no human naming required, no ambiguity for cryptic vendor filenames. The KV metadata (`origin_url`, `content_type`, `asset_slug`) serves as the lookup table. Operators can query `js-asset:{slug}` in the KV store to retrieve full provenance. The terminal summary also prints slug → origin_url at generation time.
+
+**Important:** This algorithm must produce identical output to the Proxy's KV key derivation. Engineering should implement this as a shared utility (e.g., a small JS/TS helper in the skill, or a standalone `scripts/` utility) rather than duplicating the logic.
+
+### Wildcard detection
+
+Path segments matching either pattern are replaced with `*`:
+- Semver: `\d+\.\d+[\.\d-]*` (e.g., `1.19.8-hcskhn`)
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: ❓ **question** — Slug algorithm has two ambiguities

**1. Concatenation separator:** `sha256(publisher.domain + origin_url)` — what separates the two? Raw concatenation produces `example.comhttps://vendor.io/script.js`. Since Auditor and Proxy must produce identical slugs, the separator (or its explicit absence) needs to be specified.

**2. base62 character set:** base62 is not standardized — different implementations use `[0-9A-Za-z]` vs `[0-9a-zA-Z]` vs `[A-Za-z0-9]`. The spec should pin the exact ordering or name a reference implementation.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+The original URL is preserved as a comment above the generated entry so operators can verify the wildcard substitution is correct.
+
+---
+
+## Init Mode Output
+
+### `js-assets.toml` (written to repo root)
+
+```toml
+# Generated by /audit-js-assets on 2026-04-01
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🤔 **thinking** — Hash-like regex `[A-Za-z0-9]{8,}` is too broad

The hex pattern `[a-f0-9]{6,}` is well-scoped, but `[A-Za-z0-9]{8,}` between path separators would match stable, legitimate segments like `analytics` (9), `bootstrap` (9), `modernizr` (9), `dashboard` (9). These would be incorrectly wildcarded.

Consider requiring mixed character classes (must contain both letters and digits), a higher minimum (12+), or excluding common dictionary words.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+---
+
+## Asset Entry Generation
+
+| Field | Derivation |
+|---|---|
+| `slug` | `{publisher_prefix}:{asset_stem}` — see slug algorithm below |
+| `path` | `/{publisher_prefix}/{asset_stem}.js`, or wildcard variant if versioned path detected |
+| `origin_url` | Full captured URL, with wildcard substitution applied if versioned |
+| `ttl_sec` | Omitted — proxy defaults to 1800 (wildcard) or 3600 (fixed) |
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 🤔 **thinking** — `stale_ttl_sec` is not mentioned

The Proxy spec defines `stale_ttl_sec: Option<u32>` (default 86400) in the `JsAsset` struct. This field is omitted from the Auditor's Asset Entry Generation table and generated TOML examples. While the proxy defaults handle the omission, it would be consistent to mention it (like `ttl_sec` is): _"Omitted — proxy defaults to 86400"_.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+## Implementation
+
+The Auditor is a Claude Code skill file. No compiled code.
+
+**Skill location:** `.claude/skills/audit-js-assets.md`
+
+**MCP tools used:**
+- `mcp__chrome-devtools__new_page` — open browser tab
+- `mcp__chrome-devtools__navigate_page` — load publisher URL
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: ❓ **question** — Skill location: `.claude/skills/` vs `.claude/commands/`

The spec places the skill at `.claude/skills/audit-js-assets.md`, but the project has no `.claude/skills/` directory. All existing slash commands live in `.claude/commands/` (`check-ci.md`, `verify.md`, `test-all.md`, etc.).

Is there a reason to introduce a new directory? If not, `.claude/commands/audit-js-assets.md` would be consistent with existing patterns.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
+# JS Asset Auditor — Engineering Spec
+
+**Date:** 2026-04-01  
+**Status:** Approved for engineering breakdown  
+**Related:** [JS Asset Proxy spec](2026-04-01-js-asset-proxy-design.md)
+
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: ⛏ **nitpick** — Relative link to Proxy spec may be broken at merge time

`[JS Asset Proxy spec](2026-04-01-js-asset-proxy-design.md)` assumes the Proxy spec exists in the same directory. If this PR merges first, the link is dead. Same issue on line 207.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+  xQ9pL2wY  inject_in_head=false  raven-static.vendor.io/prod/*/raven.js  [wildcard]
+  zM4nK8vP  inject_in_head=true   googletagmanager.com/gtm.js
+  ...
+
+Review inject_in_head values and commit js-assets.toml when ready.
+Diff mode: /audit-js-assets <url> --diff
+```
+
+---
+
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 👍 **praise** — Diff mode design is excellent

Appending new entries as commented-out TOML blocks keeps the file valid, requires explicit operator action, and provides full provenance. The "never auto-remove" policy for missing assets is the right default for a monitoring tool.

File: docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+1. Read `trusted-server.toml` → extract `publisher.domain` (defines first-party boundary)
+2. Open Chrome via `mcp__chrome-devtools__new_page`, navigate to target URL via `mcp__chrome-devtools__navigate_page`
+3. Wait for full page load + ~6s settle window for async script loads (`mcp__chrome-devtools__wait_for`)
+4. In parallel:
+   - `mcp__chrome-devtools__list_network_requests` → filter for requests where URL ends in `.js` or `Content-Type: application/javascript`, and origin ≠ `publisher.domain`
+   - `mcp__chrome-devtools__evaluate_script` → `Array.from(document.head.querySelectorAll('script[src]')).map(s => s.src)` → collect head-loaded script URLs
+5. Apply heuristic filter (see below)
+6. For each surviving asset, generate a `[[js_assets]]` entry (see below)
+7. Write output (init or diff mode)
```
Comment (aram356) [inline docs/superpowers/specs/2026-04-01-js-asset-auditor-design.md]: 👍 **praise** — `inject_in_head` detection via DOM query is sound

`document.head.querySelectorAll('script[src]')` correctly captures what's in `<head>` at load time, which is exactly what the Proxy needs to replicate at runtime.

**Review by aram356 — CHANGES_REQUESTED** [review]
## Summary

Well-structured engineering spec for the JS Asset Auditor. The Auditor/Proxy separation and diff mode design are strong. However, several MCP tool assumptions are incorrect and would cause the skill to fail at implementation time, and the path derivation algorithm contradicts its own examples.

## Blocking

### 🔧 wrench

- **MCP tool names are wrong**: every `mcp__chrome-devtools__*` reference uses an incorrect prefix — actual names are `mcp__plugin_chrome-devtools-mcp_chrome-devtools__*` (line 36+)
- **`wait_for` misuse**: tool waits for text on page, not a settle window — need `evaluate_script` with `setTimeout` instead (line 37)
- **`list_network_requests` filtering**: tool supports `resourceTypes` filtering, not URL/Content-Type filtering — post-processing needed (line 39)
- **Path derivation contradicts examples**: algorithm says `/{publisher_prefix}/{asset_stem}.js` but example shows `/sdk/aB3kR7mN.js` — these are different structures (line 68 vs 119)

### ❓ question

- **Slug concatenation separator**: `sha256(publisher.domain + origin_url)` doesn't specify a separator — and base62 encoding isn't standardized (character set ordering). Both must be pinned for Auditor/Proxy consistency (line 88)
- **Skill location**: spec says `.claude/skills/audit-js-assets.md` but project uses `.claude/commands/` for all slash commands — reason for new directory? (line 186)

## Non-blocking

### 🤔 thinking

- **Wildcard regex too broad**: `[A-Za-z0-9]{8,}` matches stable path segments like `analytics`, `bootstrap` (line 100)
- **`stale_ttl_sec` omitted**: Proxy spec defines this field but Auditor doesn't mention it in generated entries (line 70)
- **`platform.twitter.com`**: may need updating to `platform.x.com` post-rebrand (line 58)
- **6s settle window**: may miss late-loading ad tech scripts — consider making configurable (line 37)
- **MCP permission grants**: `close_page` and `wait_for` are not pre-approved in `.claude/settings.json` — will trigger permission prompts
- **No error handling specified**: DNS failures, ad blockers, browser crashes during sweep — worth documenting expected behavior

### 🌱 seedling

- **`docs/superpowers/specs/` is new**: existing docs use `docs/guide/`, `docs/internal/`, `docs/epics/` — convention should be documented
- **Wildcard detection asymmetry**: only defined in Auditor spec, not cross-referenced in Proxy spec — future tools won't know the patterns

### ⛏ nitpick

- **Relative link to Proxy spec**: dead if this PR merges first (line 6, 207)

### 👍 praise

- **Diff mode design**: commented-out TOML + never-auto-remove is excellent
- **Auditor/Proxy separation**: clean architectural boundary with minimal coupling
- **Terminal summary**: shows filtered-with-reason, slug mapping, wildcard indicators
- **`inject_in_head` DOM detection**: sound approach using `document.head.querySelectorAll`

## CI Status

- cargo fmt: PASS
- cargo test: PASS
- vitest: PASS
- format-typescript: PASS
- format-docs: **FAIL**
- CodeQL: PASS
- integration tests: PASS

**Comment by aram356** [comment]
Closing in place of #800
